storage/generic_mounts.go

```diff
@@ -0,0 +1,461 @@
+}
+
+func (m GenericMount) getBlock() Mount {
+	if m.Abfs != nil {
+		return m.Abfs
+	} else if m.Adl != nil {
+		return m.Adl
+	} else if m.Wasb != nil {
+		return m.Wasb
+	} else if m.S3 != nil {
+		return m.S3
+	} else if m.Gs != nil {
+		return m.Gs
+	}
+	return nil
+}
+
+// Source returns URI backing the mount
+func (m GenericMount) Source() string {
+	if block := m.getBlock(); block != nil {
+		return block.Source()
+	}
+	return m.URI
+}
+
+// Name
+func (m GenericMount) Name() string {
+	if m.MountName != "" {
+		return m.MountName
+	}
+	if block := m.getBlock(); block != nil {
+		return block.Name()
+	}
+	return ""
+}
+
+// Config returns mount configurations
+func (m GenericMount) Config(client *common.DatabricksClient) map[string]string {
+	if block := m.getBlock(); block != nil {
+		return block.Config(client)
+	}
+	return m.Options
+}
+
+// ApplyDefaults tries to apply defaults to a given resource
+func (m GenericMount) ValidateAndApplyDefaults(d *schema.ResourceData, client *common.DatabricksClient) error {
+	if block := m.getBlock(); block != nil {
+		return block.ValidateAndApplyDefaults(d, client)
+	}
+	if _, ok := d.GetOk("name"); !ok {
+		return fmt.Errorf("value of name is not specified or empty")
+	}
+	if _, ok := d.GetOk("uri"); !ok {
+		return fmt.Errorf("value of uri is not specified or empty")
+	}
+	return nil
+}
+
+// --------------- Google Cloud Storage
+
+// GSMount describes the object for a GS mount using google service account
+type GSMount struct {
+	BucketName     string `json:"bucket_name" tf:"force_new"`
+	ServiceAccount string `json:"service_account,omitempty" tf:"force_new"`
+}
+
+// Source ...
+func (m GSMount) Source() string {
+	return fmt.Sprintf("gs://%s", m.BucketName)
+}
+
+func (m GSMount) Name() string {
+	return m.BucketName
+}
+
+func (m GSMount) ValidateAndApplyDefaults(d *schema.ResourceData, client *common.DatabricksClient) error {
+	nm := d.Get("name").(string)
+	if nm != "" {
+		return nil
+	}
+	nm = m.Name()
+	if nm != "" {
+		d.Set("name", nm)
+		return nil
+	}
+	return fmt.Errorf("'name' is not detected & it's impossible to infer it")
+}
+
+// Config ...
+func (m GSMount) Config(client *common.DatabricksClient) map[string]string {
+	return make(map[string]string) // return empty map so nil map does not marshal to null
+}
+
+func preprocessGsMount(ctx context.Context, s map[string]*schema.Schema, d *schema.ResourceData, m interface{}) error {
+	var gm GenericMount
+	if err := common.DataToStructPointer(d, s, &gm); err != nil {
+		return err
+	}
+	// TODO: move into Validate function
+	if !(strings.HasPrefix(gm.URI, "gs://") || gm.Gs != nil) {
+		return nil
+	}
+	clusterID := gm.ClusterID
+	serviceAccount := ""
+	if gm.Gs != nil {
+		serviceAccount = gm.Gs.ServiceAccount
+	}
+	if clusterID == "" && serviceAccount == "" {
+		return fmt.Errorf("either cluster_id or service_account must be specified to mount GCS bucket")
+	}
+	clustersAPI := compute.NewClustersAPI(ctx, m)
+	if clusterID != "" {
```

Comment:
because we have two different cases - cluster ID is specified (we just use it), or service account is specified (we create a new cluster for it). But I rewrote the condition, so I get rid of `if` on line 138